src/lib/core/media-marshaller/media-marshaller.ts

```diff
@@ -0,0 +1,218 @@
+import {BreakPoint} from '../breakpoints/break-point';
+import {prioritySort} from '../breakpoints/breakpoint-tools';
+import {BreakPointRegistry} from '../breakpoints/break-point-registry';
+import {MatchMedia} from '../match-media/match-media';
+import {MediaChange} from '../media-change';
+
+type ValueMap = Map<string, string>;
+type BreakpointMap = Map<string, ValueMap>;
+type ElementMap = Map<HTMLElement, BreakpointMap>;
+type SubscriptionMap = Map<string, Subscription>;
+type WatcherMap = WeakMap<HTMLElement, SubscriptionMap>;
+type BuilderMap = WeakMap<HTMLElement, Map<string, Function>>;
+type SubjectMap = WeakMap<HTMLElement, Map<string, Subject<string>>>;
+
+/**
+ * MediaMarshaller - marshal MatchMedia events to their respective directives
+ */
+@Injectable({providedIn: 'root'})
+export class MediaMarshaller {
+  private activatedBreakpoints: BreakPoint[] = [];
+  private elementMap: ElementMap = new Map();
+  private watcherMap: WatcherMap = new WeakMap();
+  private builderMap: BuilderMap = new WeakMap();
+  private subjectMap: SubjectMap = new WeakMap();
+
+  get activatedBreakpoint(): string {
+    return this.activatedBreakpoints[0] ? this.activatedBreakpoints[0].alias : '';
+  }
+
+  constructor(protected matchMedia: MatchMedia,
+              protected breakpoints: BreakPointRegistry) {
+    this.matchMedia.observe().subscribe(this.activate.bind(this));
+  }
+
+  /**
+   * activate or deactivate a given breakpoint
+   * @param mc
+   */
+  activate(mc: MediaChange) {
+    const bp: BreakPoint | null = this.findByQuery(mc.mediaQuery);
+    if (mc.matches && bp) {
+      this.activatedBreakpoints.push(bp);
+      this.activatedBreakpoints.sort(prioritySort);
+    } else if (!mc.matches && bp) {
+      this.activatedBreakpoints.splice(this.activatedBreakpoints.indexOf(bp), 1);
+    }
+    this.updateStyles();
+  }
+
+  /**
+   * initialize the marshaller with necessary elements for delegation on an element
+   * @param element
+   * @param key
+   * @param builder optional so that custom bp directives don't have to re-provide this
+   * @param observables
+   */
+  init(element: HTMLElement,
+       key: string,
+       builder?: Function,
+       observables: Observable<any>[] = []): void {
+    if (builder) {
+      let builders = this.builderMap.get(element);
+      if (!builders) {
+        builders = new Map();
+        this.builderMap.set(element, builders);
+      }
+      builders.set(key, builder);
+    }
+    if (observables) {
+      let watchers = this.watcherMap.get(element);
+      if (!watchers) {
+        watchers = new Map();
+        this.watcherMap.set(element, watchers);
+      }
+      const subscription = watchers.get(key);
+      if (!subscription) {
+        const newSubscription = merge(...observables).subscribe(() => {
+          const currentValue = this.getValue(element, key);
+          this.updateElement(element, key, currentValue);
+        });
+        watchers.set(key, newSubscription);
```

Comment:
Where do we unsubscribe when the DOM element is destroyed?

Reply:
I think you've covered this in a later comment, will implement `ngOnDestroy` behavior.